invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,313 @@
+                    temp_db_conn.set_trace_callback(self._logger.debug)
+                temp_db_cursor = temp_db_conn.cursor()
+                self._run_migrations(temp_db_cursor)
+                # Close the connections, copy the original database as a backup, and move the temp database to the
+                # original database's path.
+                self._finalize_migration(
+                    temp_db_conn=temp_db_conn,
+                    temp_db_path=temp_db_path,
+                    original_db_path=self._db_path,
+                )
+            else:
+                # We are using a memory database. No special backup or special handling needed.
+                self._run_migrations(self._cursor)
+
+            self._logger.info("Database updated successfully")
+            return True
+
+    def _run_migrations(self, temp_db_cursor: sqlite3.Cursor) -> None:
+        """Runs all migrations in a loop."""
+        next_migration = self._migrations.get(from_version=self._get_current_version(temp_db_cursor))
+        while next_migration is not None:
+            self._run_migration(next_migration, temp_db_cursor)
+            next_migration = self._migrations.get(self._get_current_version(temp_db_cursor))
+
+    def _run_migration(self, migration: Migration, temp_db_cursor: sqlite3.Cursor) -> None:
+        """Runs a single migration."""
+        with self._lock:
+            try:
+                if self._get_current_version(temp_db_cursor) != migration.from_version:
+                    raise MigrationError(
+                        f"Database is at version {self._get_current_version(temp_db_cursor)}, expected {migration.from_version}"
+                    )
+                self._logger.debug(f"Running migration from {migration.from_version} to {migration.to_version}")
+
+                # Run pre-migration callbacks
+                if migration.pre_migrate:
+                    self._logger.debug(f"Running {len(migration.pre_migrate)} pre-migration callbacks")
+                    for callback in migration.pre_migrate:
+                        callback(temp_db_cursor)
+
+                # Run the actual migration
+                migration.migrate(temp_db_cursor)
+                temp_db_cursor.execute("INSERT INTO migrations (version) VALUES (?);", (migration.to_version,))
+
+                # Run post-migration callbacks
+                if migration.post_migrate:
+                    self._logger.debug(f"Running {len(migration.post_migrate)} post-migration callbacks")
+                    for callback in migration.post_migrate:
+                        callback(temp_db_cursor)
+
+                # Migration callbacks only get a cursor. Commit this migration.
+                temp_db_cursor.connection.commit()
+                self._logger.debug(
+                    f"Successfully migrated database from {migration.from_version} to {migration.to_version}"
+                )
+            except Exception as e:
```

Comment:
This is left purposefully wide - I think we want to catch everything and handle in the same way.